keps/sig-windows/4885-windows-cpu-and-memory-affinity/README.md

```diff
@@ -0,0 +1,1006 @@
+
+One difference between the Windows API and Linux is the concept of [Processor groups](https://learn.microsoft.com/windows/win32/procthread/processor-groups).
+On Windows systems with more than 64 cores the CPU's will be split into groups, 
+each processor is identified by its group number and its group-relative processor number. 
+
+In Cri we will add the following structure to the `WindowsContainerResources` in CRI:
+
+```protobuf
+message WindowsCpuGroupAffinity {
+    // CPU mask relative to this CPU group.
+    uint64 cpu_mask = 1;
+    // CPU group that this CPU belongs to.
+    uint32 cpu_group = 2;
+}
+```
+
+Since the Kubelet API's are looking for a distinct ProcessorId, the processorid's will be calculated by looping 
+through the mask and calculating the ids with `(group *64) + procesorid` resulting in unique processor id's from `group 0` as `0-63` and 
+processor Id's from `group 1` as `64-127` and so on. This translation will be done only in kubelet, the `cpu_mask` will be used when 
+communicating with the container runtime.
+
+```golang
+for i := 0; i < 64; i++ {
+		if GROUP_AFFINITY.Mask&(1<<i) != 0 {
+			processors = append(processors, i+(int(a.Group)*64))
+		}
+	}
+}
+```
+
+Using this logic, a cpu bit mask of `0000111` (leading zero's removed) would result in cpu's: 
+
+- `0,1,2` in `group 0` 
+- `64,65,66` in `group 1`.
+
+When converting back to the Windows Group Affinity we will divide the cpu number by 64 to get the group number then 
+use mod of 64 to calculate the location of the cpu in mask:
+
+```golang
+group := cpu / 64
+mask := 1 << (cpu % 64)
+groupaffinity.Mask |= mask
+```
+
+There are some scenarios where cpu count might be greater than 64 cores but in each group it is less
+than 64. For instance, you could have 2 CPU groups with 35 processors each.  The unique ID's using the strategy 
+above would give you: 
+
+- CPU group 0 : 0 to 34
+- CPU group 2: 64 to 99
+
+### Windows Memory considerations
+
+[Numa nodes](https://learn.microsoft.com/en-us/windows/win32/procthread/numa-support) can not be directly assigned or guaranteed via the Windows API but the windows sub system attempts to use memory assigned to the CPU to improve performance.  
+It is possible to indicate to a process which Numa node is preferred but a limitation of the Windows API's is that [PROC_THREAD_ATTRIBUTE_PREFERRED_NODE](https://learn.microsoft.com/windows/win32/api/processthreadsapi/nf-processthreadsapi-updateprocthreadattribute)
+does not support setting multiple Numa nodes for a single Job object (i.e. Container) so is not usable in the context of Windows containers which have multiple processes.  
+
+To work around these limitations, the kubelet will query the OS to get the affinity masks associated with each of the Numa nodes selected by the memory manager and update the CPU Group affinity accordingly in the CRI field. This will result in the memory from the Numa node being used. There are a couple scenarios that need to be considered:
```

Comment:
The intent here is to make sure if CPU's are selected that there is enough memory to also support the request.  If I understand correctly, there could be a scenario where a pod requests 4 cpus and 4gigs of memory: the cpu's might be fulfilled but the memory available on the NUMA node might only be 3gigs since another workload could be over consuming.

Reply:
(we're also discussing on sig-windows in parallel, summarizing here)
Ok, you are right that the memory manager accounts memory per-NUMA so we can still prevent some known-bad scenarios even with the more limited capabilities and guarantees windows seems to provide.
The problem I see here is this behavior, while helpful, doesn't seem to meet the guarantees the memory manager Static policy wants to provide.
I'm still reviewing docs, issues, history, code, but atm (subject to change) we may need to introduce a new policy, which linux also would need to implement, because AFAIK we don't do per-platform policies or features.

Reply:
>The problem I see here is this behavior, while helpful, doesn't seem to meet the guarantees the memory manager Static policy wants to provide.

If I understand correctly, it's because the expected guarantee on Linux is that when a NUMA node is selected by the memory manager and assigned, it can only use that node (based on Linux kernel guarantees)?   

Since the current policy is working on Windows in the prototype, in that it checks that memory is available, could we adjust the Static Policy stated guarantees  (in the docs) on Windows to be a bit softer?  

I think it's a bigger concern on Windows that we don't get the extra hint and end up doing cross node work in the scenario describe above when it could be avoided.  

> I'm still reviewing docs, issues, history, code, but atm (subject to change) we may need to introduce a new policy, which linux also would need to implement, because AFAIK we don't do per-platform policies or features.

Could a new policy be only available on Windows via it being compiled behind `_windows.go` file?  

I would have to dig deeper into the Static policy implementation but as far as I can tell right now, I don't think we would change much from Windows perspective because its providing the extra check (is there enough memory on this numa node to support the cpus selected).  The difference is what that means when it is applied on Windows vs Linux, its not 100% gauranteed since Windows doesn't have those api's to apply it that way.

Reply:
> we don't do platform-specific policies and existing policies must behave the same on supported archs

What's the reasoning?  I don't think it is realistic to have policies behave the same across OS's.  The chart above show's this is the case, `Static-> fails on startup for windows` means static is only works on Linux (assuming guarantees it makes can't be fulfilled on Windows)  

Given there are certain guarantees and expectations around `static` I can see why we don't want to use same policy name for Windows and would like to come up with a name of a new policy.  It makes sure we are giving the wrong message to end users.  I am not so sure why a new policy needs to be implemented for both OS's, especially if its not something that is not valuable and there for will not be used on Linux.

Reply:
+1 to @jsturtevant statement.

Reply:
that's my understanding, I may be wrong or misremembering. I'm not opposing this specific feature, I'm just working with that general framework in mind. Let's clarify with sig-node leads and possibly sig-arch

Reply:
> I can see a strong argument for not having `Static` work on Windows but I'm also curious why we can't have a Windows only policy or have the new policy fail at kubelet startup if targeted by a Windows node. Is there a compelling use case for the `Avoid-Known-bad (new, subject to change` policy on Linux?

Re-reading myself and the thread I realized you have a very good point @marosset !

My intention was to have as much as possible uniform behavior and same policies on both system, but my proposal indeed don't addresses this. Fundamentally, if we cannot support `Static` on windows, then we can't have the same policies on both systems, then we MUST have different policies on different systems, then there's no point in NEDING to implement the new windows-oriented policy on linux.
However, the new policy very likely can be implemented on linux as stretch goal/bonus, but by all means it won't be required.

Then, it really boils down if we can have platform-specific policies or not. My intuition was "preferably not", but like I wrote from the beginning, let's verify the requirements and the expectations here. I tried to ignite this conversation with little success, I'll try harder.

Reply:
Well, even in the case of Linux, setting the cgroup memory on a specific NUMA node doesn’t guarantee that memory allocation will always happen on that node. What is guaranteed is that the memory allocation is attempted on the specified NUMA node, but if there aren’t enough physical pages available, the Linux kernel may fail to allocate the memory, and the process can get OOM killed.

That’s why, in the memory manager’s static policy, a detailed memory consumption plan is needed for each NUMA node using an object called a memory map.

Reply:
> What is guaranteed is that the memory allocation is attempted on the specified NUMA node, but if there aren’t enough physical pages available, the Linux kernel may fail to allocate the memory, and the process can get OOM killed.

Thanks for the added details. We still won't be able to make this guarantee on Windows.  If there isn't enough memory on the NUMA node the Windows scheduler will move the memory to another Node.

Reply:
I've taken a first pass at adding information for the new Windows only Memory manager policy.  I called it `BestEffort` but that can be changed if anyone has ideas

Reply:
> Well, even in the case of Linux, setting the cgroup memory on a specific NUMA node doesn’t guarantee that memory allocation will always happen on that node. What is guaranteed is that the memory allocation is attempted on the specified NUMA node, but if there aren’t enough physical pages available, the Linux kernel may fail to allocate the memory, and the process can get OOM killed.
> 
> That’s why, in the memory manager’s static policy, a detailed memory consumption plan is needed for each NUMA node using an object called a memory map.

Thanks for keeping me honest @bg-furiosa . This distinction is important. It is still guaranteed that with Static memory manager policy if a workload runs (albeit "long enough"), then it has all its memory blocks affine to the same NUMA node.

Reply:
> I've taken a first pass at adding information for the new Windows only Memory manager policy. I called it `BestEffort` but that can be changed if anyone has ideas

`BestEffort` LGTM